### ISSUE TITLE
The totals of the cart item amounts do not match order amounts in paypal success response

### ISSUE THREAD

**Issue by shefaleechaudhary** [totaljs/node-paypal-express-checkout#24]
L_ERRORCODE0: '10413',
  L_SHORTMESSAGE0: 'Transaction refused because of an invalid argument. See additional error messages for details.',
  L_LONGMESSAGE0: 'The totals of the cart item amounts do not match order amounts.',
  L_SEVERITYCODE0: 'Error' 

Getting above error after clicking pay button of paypal page. Please help me to sort this out.

**Comment by petersirka** [comment]
@shefaleechaudhary hmm, can I see your code when you call PayPal Express Checkout?

**Comment by petersirka** [comment]
BTW: I use PayPal Express Checkout in several projects and everything works correctly.

**Comment by shefaleechaudhary** [comment]
Paypal.prototype.setExpressCheckoutPayment = function(email, invoiceNumber, amount, totalShipCost, description, currency, returnUrl, cancelUrl, onlyPayPalUsers, callback) {
	var self = this;
	var params = self.params();
	if (email) {
		params.EMAIL = email;
	}

	params.SOLUTIONTYPE = onlyPayPalUsers === true ? 'Mark' : 'Sole';
	params.PAYMENTREQUEST_0_AMT = prepareNumber(amount + totalShipCost);
	params.PAYMENTREQUEST_0_DESC = description;
	params.PAYMENTREQUEST_0_CURRENCYCODE = currency;
	params.PAYMENTREQUEST_0_INVNUM = invoiceNumber;
	params.PAYMENTREQUEST_0_CUSTOM = invoiceNumber + '|' + params.PAYMENTREQUEST_0_AMT + '|' + currency;
	params.PAYMENTREQUEST_0_PAYMENTACTION = 'Sale';
	params.PAYMENTREQUEST_0_ITEMAMT = prepareNumber(amount);
	params.PAYMENTREQUEST_0_SHIPPINGAMT = prepareNumber(totalShipCost);
 
	params = _.extend(params, this.getItemsParams());

	params.RETURNURL = returnUrl;
	params.CANCELURL = cancelUrl;

	params.NOSHIPPING = 0;
	params.ALLOWNOTE = 1;
	params.REQCONFIRMSHIPPING = 1;
	params.METHOD = 'SetExpressCheckout';

	params = _.extend(params, this.payOptions);
	
	self.request(self.url, 'POST', params, function(err, data) {
		if (err) {
			callback(err);
			return;
		}

		if (data.ACK === 'Success') {
			callback(null, { 
				redirectUrl: self.redirect + '?cmd=_express-checkout&useraction=commit&token=' + data.TOKEN, 
				token: data.TOKEN 
			});
			return;
		}

		callback(new Error('ACK ' + data.ACK + ': ' + data.L_LONGMESSAGE0));
	});

	return self;
};

**Comment by shefaleechaudhary** [comment]
@petersirka Here, I have added my code. On paypal page it shows perfect data with shipping and cart items with total. but while clicking pay button it shows error **"The totals of the cart item amounts do not match order amounts"**.

**Comment by petersirka** [comment]
Mhhh, you modified the source-code. I can't help you with this because I don't use this way for payments. Try to remove `shipping` from total amount.

**Comment by shefaleechaudhary** [comment]
Yes we have modified it. but I want to use shipping cost. Is there any other way to use shipping cost in node-paypal-express-checkout ?

**Comment by petersirka** [comment]
Just add it to total amount. I do the same.

**Comment by shefaleechaudhary** [comment]
@petersirka , But I have to show shipping cost seperately on Paypal . Any help on this from your end is highly appreciated.

**Comment by petersirka** [comment]
I understand and I can look into it within priority support <https://www.totaljs.com/support/> (around 1-3 hours), really I don't have free time for it.

Thank you!

**Comment by shefaleechaudhary** [comment]
Hello @petersirka ,
I have solved issue for shipping charges. 

Just added -
`params.PAYMENTREQUEST_0_SHIPPINGAMT = data.PAYMENTREQUEST_0_SHIPPINGAMT; `
line in Paypal.prototype.getExpressCheckoutDetails  method. 


```
Paypal.prototype.getExpressCheckoutDetails = function(token, doPayment, callback) {
	var self = this;
	var params = self.params();

	params.TOKEN = token;
	params.METHOD = 'GetExpressCheckoutDetails';

	self.request(self.url, 'POST', params, function(err, data) {
		if (err) {
			callback(err, data);
			return;
		}

		if (!doPayment) {
			return callback(null, data);
		}
		var params = self.params();
		params.PAYMENTACTION = 'Sale';
		params.PAYERID = data.PAYERID;
		params.TOKEN = token;
		params.PAYMENTREQUEST_0_AMT = data.PAYMENTREQUEST_0_AMT;
		params.PAYMENTREQUEST_0_CURRENCYCODE = data.PAYMENTREQUEST_0_CURRENCYCODE;
		params.PAYMENTREQUEST_0_ITEMAMT = data.PAYMENTREQUEST_0_ITEMAMT;
		params.PAYMENTREQUEST_0_SHIPPINGAMT = data.PAYMENTREQUEST_0_SHIPPINGAMT;
		params.METHOD = 'DoExpressCheckoutPayment';

		self.request(self.url, 'POST', params, function(err, data2) {
			if (err) {
				callback(err, data2);
				return;
			}

			if (data.ACK  !== 'Success') {
				return callback(new Error('Error DoExpressCheckoutPayment'), data2);
			}


			// Combine results of getExpressCheckout and DoExpress checkout payment.
			callback(null, _.extend(data, data2));
		});
	});

	return self;
};
```
Thank you for the help.
